File: admin/src/features/auth/repositories/auth-repository.test.ts
import {
  AxiosError,
  AxiosHeaders,
  type AxiosResponse,
  type InternalAxiosRequestConfig,
} from 'axios';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import { httpClient } from '@/core/api/http-client';
import { AppError } from '@/core/error/app-error';
import type { ApiResponse } from '@/shared/types/api-response';

import { AuthRepository } from './auth-repository';
import type { AuthResponse } from '../models/auth-response';
import type { UserResponse } from '../models/user-response';

const adminUser: UserResponse = {
  id: 1,
  email: 'admin@footverse.com',
  fullName: 'Admin',
  phone: '0900000000',
  role: 'ADMIN',
  enabled: true,
  createdAt: '2026-01-01T00:00:00',
  updatedAt: '2026-01-01T00:00:00',
};
const authData: AuthResponse = {
  accessToken: 'access-1',
  refreshToken: 'refresh-1',
  expiresIn: 900,
  tokenType: 'Bearer',
  user: adminUser,
};
const authEnvelope: ApiResponse<AuthResponse> = {
  success: true,
  message: 'OK',
  data: authData,
  timestamp: '2026-01-01T00:00:00',
};

interface CapturedRequest {
  method: string | undefined;
  url: string | undefined;
  body: unknown;
}

function okResponse<T>(config: InternalAxiosRequestConfig, data: T): AxiosResponse<T> {
  return { data, status: 200, statusText: 'OK', headers: new AxiosHeaders(), config };
}

function errorResponse(
  config: InternalAxiosRequestConfig,
  status: number,
  data: unknown,
): AxiosError {
  const response: AxiosResponse = {
    data,
    status,
    statusText: '',
    headers: new AxiosHeaders(),
    config,
  };
  return new AxiosError('Request failed', 'ERR_BAD_REQUEST', config, {}, response);
}

describe('AuthRepository', () => {
  const repository = new AuthRepository(httpClient);

  beforeEach(() => {
    localStorage.clear();
  });

  afterEach(() => {
    httpClient.defaults.adapter = undefined;
  });

  it('login issues POST /auth/login with the credentials and returns the payload', async () => {
    let captured: CapturedRequest | undefined;
    httpClient.defaults.adapter = async (config) => {
      captured = {
        method: config.method,
        url: config.url,
        body: JSON.parse(config.data as string),
      };
      return okResponse(config, authEnvelope);
    };

    const result = await repository.login({ email: 'admin@footverse.com', password: 'secret1' });

    expect(captured).toEqual({
      method: 'post',
      url: '/api/v1/auth/login',
      body: { email: 'admin@footverse.com', password: 'secret1' },
    });
    expect(result).toEqual(authData);
  });

  it('refresh issues POST /auth/refresh with the refresh token', async () => {
    let captured: CapturedRequest | undefined;
    httpClient.defaults.adapter = async (config) => {
      captured = {
        method: config.method,
        url: config.url,
        body: JSON.parse(config.data as string),
      };
      return okResponse(config, authEnvelope);
    };

    const result = await repository.refresh({ refreshToken: 'refresh-1' });

    expect(captured).toEqual({
      method: 'post',
      url: '/api/v1/auth/refresh',
      body: { refreshToken: 'refresh-1' },
    });
    expect(result).toEqual(authData);
  });

  it('logout issues POST /auth/logout with the refresh token', async () => {
    let captured: CapturedRequest | undefined;
    httpClient.defaults.adapter = async (config) => {
      captured = {
        method: config.method,
        url: config.url,
        body: JSON.parse(config.data as string),
      };
      return okResponse(config, {
        success: true,
        message: 'OK',
        data: null,
        timestamp: '2026-01-01T00:00:00',
      });
    };

    await repository.logout({ refreshToken: 'refresh-1' });

    expect(captured).toEqual({
      method: 'post',
      url: '/api/v1/auth/logout',
      body: { refreshToken: 'refresh-1' },
    });
  });

  it('surfaces an invalid-credential envelope as AppError with the errorCode intact', async () => {
    httpClient.defaults.adapter = async (config) => {
      throw errorResponse(config, 401, {
        success: false,
        message: 'Invalid email or password',
        errorCode: 'INVALID_CREDENTIALS',
        timestamp: '2026-01-01T00:00:00',
      });
    };

    const promise = repository.login({ email: 'admin@footverse.com', password: 'wrong' });

    await expect(promise).rejects.toBeInstanceOf(AppError);
    await expect(promise).rejects.toMatchObject({
      statusCode: 401,
      errorCode: 'INVALID_CREDENTIALS',
    });
  });
});

File: app/test/features/auth/repositories/auth_repository_test.dart
import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:footverse/core/error/app_exception.dart';
import 'package:footverse/features/auth/models/auth_response.dart';
import 'package:footverse/features/auth/models/forgot_password_request.dart';
import 'package:footverse/features/auth/models/login_request.dart';
import 'package:footverse/features/auth/models/refresh_token_request.dart';
import 'package:footverse/features/auth/models/register_request.dart';
import 'package:footverse/features/auth/models/reset_password_request.dart';
import 'package:footverse/features/auth/models/role.dart';
import 'package:footverse/features/auth/models/verify_reset_otp_request.dart';
import 'package:footverse/features/auth/repositories/auth_repository.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';

import 'auth_repository_test.mocks.dart';

Map<String, dynamic> _authData() => <String, dynamic>{
  'accessToken': 'access-jwt',
  'refreshToken': 'refresh-opaque',
  'expiresIn': 900,
  'tokenType': 'Bearer',
  'user': <String, dynamic>{
    'id': 1,
    'email': 'user@example.com',
    'fullName': 'Nguyen Van A',
    'phone': '0901234567',
    'avatarUrl': null,
    'role': 'CUSTOMER',
    'enabled': true,
    'createdAt': '2025-01-15T10:30:00',
    'updatedAt': '2025-01-15T10:30:00',
  },
};

Response<Map<String, dynamic>> _successResponse(String path) =>
    Response<Map<String, dynamic>>(
      requestOptions: RequestOptions(path: path),
      statusCode: 200,
      data: <String, dynamic>{
        'success': true,
        'message': 'OK',
        'data': _authData(),
        'timestamp': '2025-01-15T10:30:00',
      },
    );

DioException _errorWith(String errorCode, int statusCode) => DioException(
  requestOptions: RequestOptions(path: '/api/v1/auth'),
  error: AppException(
    message: 'error',
    statusCode: statusCode,
    errorCode: errorCode,
  ),
);

Response<void> _voidResponse(String path) =>
    Response<void>(requestOptions: RequestOptions(path: path), statusCode: 200);

Response<Map<String, dynamic>> _passwordResetTokenResponse(String path) =>
    Response<Map<String, dynamic>>(
      requestOptions: RequestOptions(path: path),
      statusCode: 200,
      data: <String, dynamic>{
        'success': true,
        'message': 'OK',
        'data': <String, dynamic>{
          'resetToken': 'opaque-reset-token',
          'expiresIn': 900,
        },
        'timestamp': '2025-01-15T10:30:00',
      },
    );

@GenerateNiceMocks([MockSpec<Dio>()])
void main() {
  late MockDio dio;
  late AuthRepository repository;

  setUp(() {
    dio = MockDio();
    repository = AuthRepository(dio);
  });

  group('register', () {
    const request = RegisterRequest(
      email: 'user@example.com',
      password: 'secret12',
      fullName: 'Nguyen Van A',
      phone: '0901234567',
    );

    test(
      'POSTs the register path with the request body, returns typed',
      () async {
        when(
          dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
        ).thenAnswer((_) async => _successResponse('/api/v1/auth/register'));

        final result = await repository.register(request);

        final captured = verify(
          dio.post<Map<String, dynamic>>(
            captureAny,
            data: captureAnyNamed('data'),
          ),
        ).captured;
        expect(captured[0], '/api/v1/auth/register');
        expect(captured[1], {
          'email': 'user@example.com',
          'password': 'secret12',
          'fullName': 'Nguyen Van A',
          'phone': '0901234567',
        });
        expect(result, isA<AuthResponse>());
        expect(result.accessToken, 'access-jwt');
        expect(result.user.role, Role.customer);
      },
    );

    test('surfaces USER_EMAIL_DUPLICATED as AppException', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('USER_EMAIL_DUPLICATED', 409));

      await expectLater(
        repository.register(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'USER_EMAIL_DUPLICATED',
          ),
        ),
      );
    });
  });

  group('login', () {
    const request = LoginRequest(
      email: 'user@example.com',
      password: 'secret12',
    );

    test('POSTs the login path with the request body, returns typed', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenAnswer((_) async => _successResponse('/api/v1/auth/login'));

      final result = await repository.login(request);

      final captured = verify(
        dio.post<Map<String, dynamic>>(
          captureAny,
          data: captureAnyNamed('data'),
        ),
      ).captured;
      expect(captured[0], '/api/v1/auth/login');
      expect(captured[1], {
        'email': 'user@example.com',
        'password': 'secret12',
      });
      expect(result.accessToken, 'access-jwt');
      expect(result.refreshToken, 'refresh-opaque');
      expect(result.expiresIn, 900);
    });

    test('surfaces INVALID_CREDENTIALS as AppException', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('INVALID_CREDENTIALS', 401));

      await expectLater(
        repository.login(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'INVALID_CREDENTIALS',
          ),
        ),
      );
    });
  });

  group('refresh', () {
    const request = RefreshTokenRequest(refreshToken: 'refresh-opaque');

    test(
      'POSTs the refresh path with the request body, returns typed',
      () async {
        when(
          dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
        ).thenAnswer((_) async => _successResponse('/api/v1/auth/refresh'));

        final result = await repository.refresh(request);

        final captured = verify(
          dio.post<Map<String, dynamic>>(
            captureAny,
            data: captureAnyNamed('data'),
          ),
        ).captured;
        expect(captured[0], '/api/v1/auth/refresh');
        expect(captured[1], {'refreshToken': 'refresh-opaque'});
        expect(result, isA<AuthResponse>());
      },
    );

    test('surfaces REFRESH_TOKEN_INVALID as AppException', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('REFRESH_TOKEN_INVALID', 401));

      await expectLater(
        repository.refresh(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'REFRESH_TOKEN_INVALID',
          ),
        ),
      );
    });
  });

  group('logout', () {
    const request = RefreshTokenRequest(refreshToken: 'refresh-opaque');

    test('POSTs the logout path with the refresh token body', () async {
      when(dio.post<void>(any, data: anyNamed('data'))).thenAnswer(
        (_) async => Response<void>(
          requestOptions: RequestOptions(path: '/api/v1/auth/logout'),
          statusCode: 200,
        ),
      );

      await repository.logout(request);

      final captured = verify(
        dio.post<void>(captureAny, data: captureAnyNamed('data')),
      ).captured;
      expect(captured[0], '/api/v1/auth/logout');
      expect(captured[1], {'refreshToken': 'refresh-opaque'});
    });
  });

  group('forgotPassword', () {
    const request = ForgotPasswordRequest(email: 'user@example.com');

    test('POSTs the forgot-password path with the request body', () async {
      when(
        dio.post<void>(any, data: anyNamed('data')),
      ).thenAnswer((_) async => _voidResponse('/api/v1/auth/forgot-password'));

      await repository.forgotPassword(request);

      final captured = verify(
        dio.post<void>(captureAny, data: captureAnyNamed('data')),
      ).captured;
      expect(captured[0], '/api/v1/auth/forgot-password');
      expect(captured[1], {'email': 'user@example.com'});
    });
  });

  group('verifyResetOtp', () {
    const request = VerifyResetOtpRequest(
      email: 'user@example.com',
      otp: '123456',
    );

    test(
      'POSTs the verify-reset-otp path with the request body, returns typed',
      () async {
        when(
          dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
        ).thenAnswer(
          (_) async =>
              _passwordResetTokenResponse('/api/v1/auth/verify-reset-otp'),
        );

        final result = await repository.verifyResetOtp(request);

        final captured = verify(
          dio.post<Map<String, dynamic>>(
            captureAny,
            data: captureAnyNamed('data'),
          ),
        ).captured;
        expect(captured[0], '/api/v1/auth/verify-reset-otp');
        expect(captured[1], {'email': 'user@example.com', 'otp': '123456'});
        expect(result.resetToken, 'opaque-reset-token');
        expect(result.expiresIn, 900);
      },
    );

    test('surfaces PASSWORD_RESET_OTP_INVALID as AppException', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('PASSWORD_RESET_OTP_INVALID', 400));

      await expectLater(
        repository.verifyResetOtp(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'PASSWORD_RESET_OTP_INVALID',
          ),
        ),
      );
    });

    test('surfaces PASSWORD_RESET_OTP_EXPIRED as AppException', () async {
      when(
        dio.post<Map<String, dynamic>>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('PASSWORD_RESET_OTP_EXPIRED', 400));

      await expectLater(
        repository.verifyResetOtp(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'PASSWORD_RESET_OTP_EXPIRED',
          ),
        ),
      );
    });
  });

  group('resetPassword', () {
    const request = ResetPasswordRequest(
      resetToken: 'opaque-reset-token',
      newPassword: 'NewPass123',
    );

    test('POSTs the reset-password path with the request body', () async {
      when(
        dio.post<void>(any, data: anyNamed('data')),
      ).thenAnswer((_) async => _voidResponse('/api/v1/auth/reset-password'));

      await repository.resetPassword(request);

      final captured = verify(
        dio.post<void>(captureAny, data: captureAnyNamed('data')),
      ).captured;
      expect(captured[0], '/api/v1/auth/reset-password');
      expect(captured[1], {
        'resetToken': 'opaque-reset-token',
        'newPassword': 'NewPass123',
      });
    });

    test('surfaces PASSWORD_RESET_TOKEN_INVALID as AppException', () async {
      when(
        dio.post<void>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('PASSWORD_RESET_TOKEN_INVALID', 400));

      await expectLater(
        repository.resetPassword(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'PASSWORD_RESET_TOKEN_INVALID',
          ),
        ),
      );
    });

    test('surfaces PASSWORD_RESET_TOKEN_EXPIRED as AppException', () async {
      when(
        dio.post<void>(any, data: anyNamed('data')),
      ).thenThrow(_errorWith('PASSWORD_RESET_TOKEN_EXPIRED', 400));

      await expectLater(
        repository.resetPassword(request),
        throwsA(
          isA<AppException>().having(
            (e) => e.errorCode,
            'errorCode',
            'PASSWORD_RESET_TOKEN_EXPIRED',
          ),
        ),
      );
    });
  });
}
